### PR TITLE
Add groupby shift method

### LINKED ISSUE #7095: Groupby shift
I'd like to be able to use `shift` on groupby Series and DataFrame objects to create per-group lag or lead columns. Today, I can do this in pandas but not Dask.

```python
import pandas as pd
import dask.dataframe as dd
​
pdf = pd.DataFrame({
    "a": [0,1,0,1,1,0],
    "b": range(6),
    "c": ["a","b","c","d","e","f"]
})
ddf = dd.from_pandas(pdf, 2)
​
print(pdf.groupby("a").b.shift(1))
ddf.groupby("a").b.shift(1)
0    NaN
1    NaN
2    0.0
3    1.0
4    3.0
5    2.0
Name: b, dtype: float64
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-17-f89d506bbdcd> in <module>
     11 
     12 print(pdf.groupby("a").b.shift(1))
---> 13 ddf.groupby("a").b.shift(1)

AttributeError: 'SeriesGroupBy' object has no attribute 'shift'
```

```
conda list | grep "pandas\|dask"
dask                      2021.1.0           pyhd8ed1ab_0    conda-forge
pandas                    1.1.5            py37hdc94413_0    conda-forge
```

### PR DESCRIPTION
**Pull request by kori73**
- [x] Closes #7095
- [x] Tests added / passed
- [x] Passes `pre-commit run --all-files`

I have to implemented the shift `method` following the `transform` and `apply` methods. Let me know if you'd like any changes/improvements.

### CODE REVIEW

**Comment by GPUtester** [comment]
Can one of the admins verify this patch?

**Comment by jsignell** [comment]
ok to test

**Comment by jsignell** [comment]
rerun tests

File: dask/dataframe/groupby.py [hunk 1]
```diff
@@ -1800,6 +1813,94 @@ def transform(self, func, *args, **kwargs):
...
+            warnings.warn(msg, stacklevel=2)
+
+        meta = make_meta(meta, parent_meta=self._meta.obj)
+
+        # Validate self.by
+        if isinstance(self.by, list) and any(
+            isinstance(item, Series) for item in self.by
+        ):
+            raise NotImplementedError(
+                "groupby-transform with a multiple Series is currently not supported"
```
Comment (jsignell) [inline dask/dataframe/groupby.py:1876]: copy-paste holdover :)
```suggestion
                "groupby-shift with a multiple Series is currently not supported"
```

File: dask/dataframe/groupby.py [hunk 2]
```diff
@@ -1800,6 +1813,94 @@ def transform(self, func, *args, **kwargs):
...
+        # Perform embarrassingly parallel groupby-shift
+        result = map_partitions(
+            _groupby_slice_shift,
+            df2,
+            by,
+            self._slice,
+            periods=periods,
+            freq=freq,
+            axis=axis,
+            fill_value=fill_value,
```
Comment (jsignell) [inline dask/dataframe/groupby.py:1896]: Probably best to explicitly include the token for groupby-shift like we do for `apply` and `transform`.
```suggestion
            fill_value=fill_value,
            token="groupby-shift",
```

**Review by jsignell — COMMENTED** [review]
This looks really good @kori73 thanks for doing this work! I just have a few suggestions.

**Comment by jsignell** [comment]
This is in @kori73! Thanks for doing this work :)

**Pull request merged**